### PR TITLE
Setup

### PR DESCRIPTION
**Pull request by grumpy**
Allows basic editing - 
While a shape is selected with mouse down, Plus and Minus changed the z order and Delete / Backspace deleted the shape.

Known deficiencies -
I didn't any styes in css.
./redux/app.ts is a little dense, it has everything related to the state and reducers.
I commented on a few things inline that I'm not sure about.

### CODE REVIEW

File: src/components/DragLayer.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import {connect} from "react-redux";
+import {Dispatch} from "redux";
+import * as React from "react";
+import {App, dragShape, dropShape, moveShape, deselectShape,
+    moveShapeForward, moveShapeBackward, deleteShape} from "../redux/app.ts";
+import * as shape from "./Shape.tsx";
+
+export class DragLayer extends React.Component<App & any, void> {
+  translatePoint(x: number, y: number) {
+    const canvas = document.getElementById("drawingCanvas");
```
Comment (grumpy) [inline src/components/DragLayer.tsx]: This doesn't seem very React flavored to me, but I couldn't get any alternatives to work.  The ref wasn't accessible from here.

File: src/components/DragLayer.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        this.props.dispatch(moveShapeBackward());
+      }
+      else if (e.key == "Backspace" || e.key == "Delete") {
+        this.props.dispatch(deleteShape());
+      }
+    }
+  }
+
+  componentDidMount() {
+    window.onkeydown = this.handleKeyDown.bind(this);
```
Comment (grumpy) [inline src/components/DragLayer.tsx]: This doesn't seem very React flavored to me, but I couldn't get the react keyboard events to work.

**Comment by grumpy** [comment]
@danschultz Working version.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    "url": "https://github.com/grumpy/learn_typescript/issues"
+  },
+  "homepage": "https://github.com/grumpy/learn_typescript#readme",
+  "dependencies": {
+    "react": "^15.2.1",
+    "react-dnd": "^2.1.4",
+    "react-dom": "^15.2.1",
+    "react-redux": "^4.4.5",
+    "redux": "^3.5.2",
+    "redux-devtools": "^3.3.1"
```
Comment (danschultz) [inline package.json]: this should be a dev dependency.

File: src/components/DragLayer.tsx (hunk 1 above)
Comment (danschultz) [inline src/components/DragLayer.tsx]: Would be curious to see how you were using refs. It should work.

File: src/components/DragLayer.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  render(): JSX.Element {
+    if (this.props.drag != null) {
+      const halfWidth: number = this.props.drag.width / 2.0;
+      const halfHeight: number = this.props.drag.height / 2.0;
+
+      return (
+        <div style={{ outline: "none", cursor: "pointer",
+            overflow: "hidden", position: "absolute", top: 0, left: 0, width: "100%", height: "100%" }}
+          onMouseMove={this.handleMove.bind(this)}
+          onMouseUp={this.handleUp.bind(this)}>
```
Comment (danschultz) [inline src/components/DragLayer.tsx]: Would recommend adding your `onMouseMove` and `onMouseUp` handlers to window in your `onMouseDown` handler. This way the dragging won't stop if the user moves their mouse outside the element.

File: src/components/DrawingCanvas.tsx [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import {App, Shape, selectShape} from "../redux/app.ts";
+import * as shape from "./Shape.tsx";
+
+export class DrawingCanvas extends React.Component<App & any, void> {
+  renderShape(shape: Shape, index: number): JSX.Element {
+    const color: string = (this.props.selected == index) ? "rgb(0,0,0)" : null;
+    const x: number = shape.x - (shape.size / 2.0);
+    const y: number = shape.y - (shape.size / 2.0);
+    return (
+      <ConnectedShape key={"shape"+index} index={index} shape={shape.shape} color={color} size={100} x={x} y={y} />
```
Comment (danschultz) [inline src/components/DrawingCanvas.tsx]: Not a big deal in this case, but the index should be tied to the shape data. During reordering, the keys won't remap to the shape that the component is rendering.

File: src/components/ShapesPalette.tsx [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+          </svg>
+        </div>
+        <div style={{ margin: "5vh auto", width: "10vh", height: "10vh" }}>
+          <svg viewBox="0 0 100 100" style={{ width: "100%", height: "100%" }}>
+            <ConnectedShape shape={"circle"} size={100}/>
+          </svg>
+        </div>
+        <div style={{ margin: "5vh auto", width: "10vh", height: "10vh" }}>
+          <svg viewBox="0 0 100 100" style={{ width: "100%", height: "100%" }}>
+            <ConnectedShape shape={"triangle"} size={100}/>
```
Comment (danschultz) [inline src/components/ShapesPalette.tsx]: 👍  on reusing shape drawing

File: src/redux/app.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+    {
+      const dragShape: DragShapeAction = action as DragShapeAction;
+      const drag:Drag = Object.assign({}, state.drag, {x: dragShape.x, y:dragShape.y});
+      return Object.assign({}, state, {drag: drag});
+    } 
+    case DROP_SHAPE:
+    {
+      const dropShape: DropShapeAction = action as DropShapeAction;
+      const shapes = state.shapes.slice();
+      shapes.push({shape: state.drag.shape, size: state.drag.size, x: dropShape.x, y: dropShape.y});
```
Comment (danschultz) [inline src/redux/app.ts]: Tip: You can use destructuring to create new arrays.

```ts
const newShape = {shape: state.drag.shape, size: state.drag.size, x: dropShape.x, y: dropShape.y}
const shapes = [...state.shapes, newShape];
```

File: src/components/DragLayer.tsx (hunk 2 above)
Comment (danschultz) [inline src/components/DragLayer.tsx]: Hmm, AFAIU all elements should get keyboard events.

**Comment by danschultz** [comment]
LGTM

**Comment by danschultz** [comment]
LGTM

**Pull request closed without merging**